tools/img_split_dota.py

```diff
@@ -0,0 +1,363 @@
+
+    window_anns = []
+    for i in range(windows.shape[0]):
+        win_iofs = iofs[:, i]
+        pos_inds = np.nonzero(win_iofs >= iof_thr)[0].tolist()
+
+        win_ann = dict()
+        for k, v in info['ann'].items():
+            try:
+                win_ann[k] = v[pos_inds]
+            except TypeError:
+                win_ann[k] = [v[i] for i in pos_inds]
+        win_ann['trunc'] = win_iofs[pos_inds] < 1
+        window_anns.append(win_ann)
+    return window_anns
+
+
+def crop_and_save_img(info, windows, window_anns, img_dir, no_padding,
+                      padding_value, filter_empty, save_dir, img_ext, generate_txt, save_anns):
+    img = cv2.imread(osp.join(img_dir, info['filename']))
+    patch_infos = []
+    for i in range(windows.shape[0]):
+        ann = window_anns[i]
+        if filter_empty and (ann['bboxes'].size == 0):
+            continue
+
+        patch_info = dict()
+        for k, v in info.items():
+            if k not in ['id', 'fileanme', 'width', 'height', 'ann']:
+                patch_info[k] = v
+
+        window = windows[i]
+        x_start, y_start, x_stop, y_stop = window.tolist()
+        ann['bboxes'] = bt.translate(ann['bboxes'], -x_start, -y_start)
+        patch_info['ann'] = ann
+        patch_info['x_start'] = x_start
+        patch_info['y_start'] = y_start
+        patch_info['id'] = info['id'] + f'_{i:04d}'
+        patch_info['ori_id'] = info['id']
+
+        patch = img[y_start:y_stop, x_start:x_stop]
+        if not no_padding:
+            height = y_stop - y_start
+            width = x_stop - x_start
+            if height > patch.shape[0] or width > patch.shape[1]:
+                padding_patch = np.empty(
+                    (height, width, patch.shape[-1]), dtype=np.uint8)
+                if not isinstance(padding_value, (int, float)):
+                    assert len(padding_value) == patch.shape[-1]
+                padding_patch[...] = padding_value
+                padding_patch[:patch.shape[0], :patch.shape[1], ...] = patch
+                patch = padding_patch
+        patch_info['height'] = patch.shape[0]
+        patch_info['width'] = patch.shape[1]
+
+        cv2.imwrite(osp.join(save_dir, patch_info['id']+img_ext), patch)
+        patch_info['filename'] = patch_info['id'] + img_ext
+        patch_infos.append(patch_info)
+        
+        if generate_txt:
+            #annotation txt file generation
+            txtpath = osp.join(save_anns, patch_info['id']+'.txt')
+            txtfile = open(txtpath, "a");
+            for i in range(len(patch_info['ann']['labels'])):
+                box = np.array2string(patch_info['ann']['bboxes'][i]).strip("[""]")
+                box += ' '
+                box += CLASSES[patch_info['ann']['labels'][i]]
+                box += ' '
+                box += str(patch_info['ann']['diffs'][i])
+                box += '\n'
+                #print(box)
+                txtfile.write(box)
+            txtfile.close()
```

Comment:
It's better to put txt files into a new folder. You can create a new `save_txt` folder by `save_anns = osp.join(args.save_dir, 'txt')`

Reply:
ok